server/datastore/mysql/software.go

```diff
@@ -517,16 +476,112 @@ func insertNewInstalledHostSoftwareDB(
+				return nil, ctxerr.New(ctx, fmt.Sprintf("software not found for checksum %s", hex.EncodeToString([]byte(s.Checksum))))
+			}
+			software.ID = s.ID
+			insertsHostSoftware = append(insertsHostSoftware, hostID, software.ID, software.LastOpenedAt)
+			insertedSoftware = append(insertedSoftware, *software)
+			delete(softwareChecksums, s.Checksum)
+		}
+	}
+
+	// For software items that don't already exist in the software table, we insert them.
+	if len(softwareChecksums) > 0 {
+		keys := make([]string, 0, len(softwareChecksums))
+		for checksum := range softwareChecksums {
+			keys = append(keys, checksum)
+		}
+		for i := 0; i < len(keys); i += softwareInsertBatchSize {
+			start := i
+			end := i + softwareInsertBatchSize
+			if end > len(keys) {
+				end = len(keys)
+			}
+			totalToProcess := end - start
+			const numberOfArgsPerSoftware = 10 // number of ? in each VALUES clause
+			values := strings.TrimSuffix(
+				strings.Repeat("(?,?,?,?,?,?,?,?,?,?),", totalToProcess), ",",
+			)
+			// INSERT IGNORE is used to avoid duplicate key errors, which may occur since our previous read came from the replica.
+			stmt := fmt.Sprintf(
+				"INSERT IGNORE INTO software (name, version, source, `release`, vendor, arch, bundle_identifier, extension_id, browser, checksum) VALUES %s",
+				values,
+			)
+			args := make([]interface{}, 0, totalToProcess*numberOfArgsPerSoftware)
+			for j := start; j < end; j++ {
+				checksum := keys[j]
+				sw := softwareChecksums[checksum]
+				args = append(
+					args, sw.Name, sw.Version, sw.Source, sw.Release, sw.Vendor, sw.Arch, sw.BundleIdentifier, sw.ExtensionID, sw.Browser,
+					checksum,
+				)
+			}
+			if _, err := tx.ExecContext(ctx, stmt, args...); err != nil {
+				return nil, ctxerr.Wrap(ctx, err, "insert software")
+			}
+		}
+
+		// Here, we use the transaction (tx) for retrieval because we must retrieve the software IDs that we just inserted.
+		existingSoftware, err := getSoftwareIDsByChecksums(ctx, tx, keys)
+		if err != nil {
+			return nil, err
+		}
+		for _, s := range existingSoftware {
+			software, ok := softwareChecksums[s.Checksum]
+			if !ok {
+				return nil, ctxerr.New(ctx, fmt.Sprintf("software not found for checksum %s", hex.EncodeToString([]byte(s.Checksum))))
+			}
+			software.ID = s.ID
+			insertsHostSoftware = append(insertsHostSoftware, hostID, software.ID, software.LastOpenedAt)
+			insertedSoftware = append(insertedSoftware, *software)
+			delete(softwareChecksums, s.Checksum)
+		}
+	}
+
+	if len(softwareChecksums) > 0 {
+		// We log and continue. We should almost never see this error. If we see it regularly, we need to investigate.
+		level.Error(ds.logger).Log(
+			"msg", "could not find or create software items. This error may be caused by master and replica DBs out of sync.", "host_id",
+			hostID, "number", len(softwareChecksums),
+		)
+		for checksum, software := range softwareChecksums {
+			uuidString := ""
+			checksumAsUUID, err := uuid.FromBytes([]byte(checksum))
```

Comment:
Curious about why we need to convert to UUID (non printable chars?)

Reply:
This is just for debugging. In MySQL, we can use [BIN_TO_UUID](https://dev.mysql.com/doc/refman/8.0/en/miscellaneous-functions.html#function_bin-to-uuid) to compare to the printed value.

Reply:
I don't expect these print statements to be hit.